lib/buffy/throttle_and_timed.ex

```diff
@@ -0,0 +1,387 @@
+      that is based on [GenServer's timeout feature](https://hexdocs.pm/elixir/1.12/GenServer.html#module-timeouts).
+
+  Main reason for these changes is sometimes there's a need to fall back to a time-interval triggered work, when there aren't any triggers to
+  start the work. Requirement of this means the process should exist and not get terminated immediately after a successfully throttled work execution.
+
+  ### In other words, we keep the throttle mechanism:
+
+  Once the timer has expired, the function will be called,
+  and any subsequent calls will start a new timer.
+
+  ```text
+  call     call   call               call           call
+   | call   | call | call             | call         |
+   |  |     |  |   |  |               |  |           |
+  ┌─────────┐  ┌─────────┐            ┌─────────┐    ┌─────────┐
+  │ Timer 1 │  │ Timer 2 │            │ Timer 3 │    │ Timer 4 │
+  └─────────|  └─────────┘            └─────────┘    └─────────┘
+            |            |                      |              |
+            |            |                      |    Forth function invocation
+            |            |            Third function invocation
+            | Second function invocation
+  First function invocation
+  ```
+
+  ### With the optionally enabled trigger, ending up in this lifecycle:
+
+  ```mermaid
+  graph TB
+    A[Start Buffy] -->|start_link| B(Init Buffy)
+    B --> |initial handle_continue| W(Do throttled work)
+    S(Messages sent to Buffy) --> |message to trigger work| D{Can Throttle?}
+    D --> |YES| W
+    D --> |NO| C(Ignore message as throttle already scheduled)
+    S --> |empty inbox timeout interval| P(Do immediate work)
+    W --> |set message inbox timeout| S
+    P --> |set message inbox timeout| S
+  ```
+
+  ### Note on Horde based usage
+
+  Under Horde, the state unfortunately doesn't get synced up automatically - that requires explicit tooling.
+  Therefore state will be "reset" to the initial state when process boots up. This is not a big issue as the initial state is to
+  set a throttled run of `handle_throttle`.
+
+  ### How to start timed interval triggers when your application boots up
+
+  This will not run when your application starts. The suggested approach is to create a child spec for the application Supervisor (typically in `application.ex`)
+  for a Task module, that runs how many instances of `throttle/1` as necessary. That way, the default inbox timeout will run,
+  using `loop_interval`, as some variation of:
+
+  ```
+  # application.ex
+  def start(_type, _args) do
+    ...
+    children = [
+      ...
+      {true,
+       Supervisor.child_spec(
+         {Task,
+          fn ->
+            for x <- 1..10, do: MyModuleUsingThrottleAndTimed.throttle(some: "value", x: x)
+          end},
+         id: MyModuleUsingThrottleAndTimedInit,
+         restart: :temporary
+       )}
+    ]
+    ...
+  ```
```

Comment:
Is this how we plan to use this on the wms? I was envisioning that the servers would _start_ on demand but just keep running thereafter. That way, if we want to start a new process for a new shipper (in the wms case), we don't need to couple that with an application re-start to create a process for that shipper.

Reply:
Agreed - and that's how it is implemented. What was added here is the initial runs to cover the scenario of low-volume shippers which won't have the initial "start on demand" event. We want to make the loop interval start regardless of an initial event.

Reply:
As discussed we will start off without this change in `wms-service` - we don't want to increase chance of overwhelming the DB reads during deployment when it's possibly not necessary.